File: runtime/python/angine/pdp.py
import lupa
from typing import List
from .handlers import *
from .context import ResponseCtx, EvaluationCtx
from .results import Decision, Status, Result


class PDP:
    LUA_ENTRY = "__main"

    def __init__(self, lua_policy: str, handlers=None):
        """
        :param lua_policy: generated Lua policy
        :param handlers: instance of a class containing access control helpers
        """
        self.handlers = handlers
        lua = lupa.LuaRuntime(
            attribute_handlers=(getter, setter),
            unpack_returned_tuples=True
        )
        if lua is None:
            raise RuntimeError("Lua runtime can't be initialized")

        self.lua_runtime = lua
        self.actions = self.lua_runtime.table_from({
            e.name.lower(): e.value for e in Decision
        })

        # Look for the policy entry point
        lua.execute(lua_policy)
        g = lua.globals()
        self.lua_policy = None
        for gname, gfunc in dict(g).items():
            if gname.startswith(self.LUA_ENTRY):
                self.lua_policy = gfunc
                break
        else:
            raise ValueError("ALFAScript policy entry point not found")

    def evaluate(self, evaluation_ctx: EvaluationCtx) -> ResponseCtx:
        """ Evaluates the @self.lua_policy for all given contexts

        :param evaluation_ctx: context, containing all necessary information for evaluation
        :return if @format_results to True returns list of decisions in a format
            described in `self.get_response` otherwise returns list of Decision enums
        """
        ctx_list = evaluation_ctx.access_requests

        results = [
            Result(
                Decision(
                    self.lua_policy(ctx, self.actions, self.handlers)
                )
            ) for ctx in ctx_list
        ]
        if evaluation_ctx.combined_decision:
            return ResponseCtx(
                [PDP.get_xacml3_combined_decision(results)]
            )
        else:
            return ResponseCtx(results)

    @staticmethod
    def get_xacml3_combined_decision(results: List[Result]) -> Result:
        """
        Implements XACML3 Multiple Decision Profile combining method.
        """
        if all(result.decision == results[0].decision for result in results):
            return results[0]
        else:
            return Result(
                Decision.Indeterminate, Status(Status.Code.ProcessingError)
            )

File: runtime/java/angine/src/main/java/angine/PDP.java
package angine;

import angine.context.AccessRequest;
import angine.context.EvaluationContext;
import angine.context.ResponseContext;
import angine.result.Result;
import org.luaj.vm2.Globals;
import org.luaj.vm2.LuaValue;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;
import org.luaj.vm2.lib.jse.JsePlatform;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class PDP {

    public static int PERMIT = 0;
    public static int DENY = 1;
    public static int NOT_APPLICABLE = 2;
    public static int INDETERMINATE = 3;




    private String LUA_ENTRY = "__main";

    private LuaValue actions;

    private LuaValue luaPolicy;

    private Object handlers;

    public PDP(String policy){
        this(policy, null);
    }

    public PDP(String policy, @Nullable Object handlers) throws IllegalArgumentException{
        this.actions = LuaValue.tableOf();
        actions.set("permit", PERMIT);
        actions.set("deny", DENY);
        actions.set("notapplicable", NOT_APPLICABLE);
        actions.set("indeterminate", INDETERMINATE);

        Globals globals = JsePlatform.standardGlobals();
        globals.set("actions",actions);
        globals.load(policy).call();
        this.luaPolicy = globals.get(LUA_ENTRY);
        if(luaPolicy == null){
            throw new IllegalArgumentException("Policy's entry point not found.");
        }
        this.handlers = handlers;
    }


    public ResponseContext evaluate(EvaluationContext evaluationContext){
        List<AccessRequest> accessRequests = evaluationContext.accessRequests;
        List<Result> results = new ArrayList<Result>();
        for(AccessRequest accessRequest : accessRequests){
            LuaValue ctx = CoerceJavaToLua.coerce(accessRequest);
            LuaValue handlers = CoerceJavaToLua.coerce(this.handlers);
            LuaValue res = luaPolicy.call(ctx, actions, handlers);
            results.add( new Result(
                    Decision.fromInt(res.toint())
            ));
        }
        if(evaluationContext.combinedDecision){
            return new ResponseContext(Collections.singletonList(getCombinedDecision(results)));
        } else {
            return new ResponseContext(results);
        }

    }


    public static Result getCombinedDecision( List<Result> results){
        if(results == null || results.size() == 0){
            return new Result(
                    Decision.Indeterminate,
                    new Status(Status.Code.PROCESSING_ERROR)
            );
        }

        for(Result result : results){
            if(results.get(0).decision != result.decision){
                return new Result(
                        Decision.Indeterminate,
                        new Status(Status.Code.PROCESSING_ERROR)
                );
            }
        }
        return results.get(0);

    }


}